src/Storages/MergeTree/MergeTreeSink.cpp

```diff
@@ -17,29 +17,72 @@ void MergeTreeSink::onStart()
 
 void MergeTreeSink::consume(Chunk chunk)
 {
-    auto block = getHeader().cloneWithColumns(chunk.detachColumns());
+    const auto settings = storage.getSettings();
+    std::unique_ptr<ThreadPool> pool = nullptr;
+    if (part_blocks.size() > 1 && settings->max_part_writing_threads > 1)
```

Comment:
But most of the time, likely, there will be only one part, right?
FWIW there was some similar patches, #16503/#14150 and more recent for s3 only - #33291

Reply:
I am not sure about it, but in our scenario, there are usually 5 to 8 parts in one block. I think it also depends on some other settings, such as with or without partition key, max_parts_per_block, xxx_block_size, etc.
thanks, azat, These patches look much more complicated, I'll take a look on these.